SPARK review

jklmnn opened the pull request:
DO NOT MERGE, this pull request is only for code review.

jklmnn commented on src/lib/cbe/cbe-primitive.ads:
```diff
@@ -0,0 +1,94 @@
+	use type Request.Success_Type;
+	use type Request.Block_Number_Type;
+
+	type Index_Type  is mod 2**64;
```
This declaration hides `CBE.Index_Type`. You might either rename it or even drop this declaration completely.

jklmnn commented on src/server/aes_cbc_block/cbe-cxx-cxx_pool.adb:
```diff
@@ -0,0 +1,130 @@
+			raise Program_Error;
+		elsif not Correct_Object_Size(Prim'Size, Prim_Size) then
+			raise Program_Error;
+		end if;
```
```Ada
if
   not Fitting_Object_Size(Obj'Size, Obj_Size)
   or else not Correct_Object_Size(Req'Size, Req_Size)
   or else not Correct_Object_Size(Prim'Size, Prim_Size)
then
   raise Program_Error;
end if;
```
Is equivalent to this expression. Again this should be handled in a precondition to prove that the exceptions will never be raised.

jklmnn commented on src/server/aes_cbc_block/cbe-cxx-cxx_splitter.adb:
```diff
@@ -0,0 +1,108 @@
+		Req  : constant CXX_Request.Object_Type   := (0,0,0,0,0,0);
+		Prim : constant CXX_Primitive.Object_Type := (0,0,0,0,0);
+	begin
+		if not Correct_Object_Size(Obj'Size, Obj_Size) then
```
Same as above.

jklmnn commented on src/server/aes_cbc_block/cbe-cxx-cxx_splitter.adb:
```diff
@@ -0,0 +1,108 @@
+		if not Primitive.Valid(Spark_Prim) then
+			raise Program_Error;
+		end if;
+		if Primitive."/="(Splitter.Curr_Primitive(Obj), Spark_Prim) then
```
You could also add a
```Ada
use type Primitive;
```
in the declaration block of the procedure to make the type operators available and then write
```Ada
if Splitter.Curr_Primitive(Obj) /= Spark_Prim then
```
Same accounts for other occurrences of this.

jklmnn commented on src/server/aes_cbc_block/cbe-pool.adb:
```diff
@@ -0,0 +1,263 @@
+				Request.Success(Rq, False);
+				Item.Req(Obj, Rq);
+			end if;
+			Obj.Nr_Of_Done_Prims := Obj.Nr_Of_Done_Prims + 1;
```
Is an overflow in this addition desirable? If not you should change `Number_Of_Primitives_Type` to a ranged type (and then check that no overflow can occur).

jklmnn reviewed (commented):
@cnuke
I added some comments to your code, here also some general things:

When raising an exception you still should add a precondition to the procedure that makes sure that the exception is never raised as long as the precondition is met.

Also many procedure made some checks to return if those weren't met. These are good points to remove those checks and add then as precondition so the calling code needs to check (where you then also could add preconditions e.g.). The idea is to handle errors as early as possible.

At some points you return records/structs. I'm not sure about the exact calling convention here. Maybe you should add `pragma Convention (C_Pass_By_Copy, Typename);` to the binding types if you pass them by copy as Ada will choose the calling convention depending on the type size and some other properties.

cnuke commented:
@jklmnn thanks a lot for the review and the helpful suggestions. I will apply them when cleaning up the branch.

(I am sorry for the late reply, I just saw it yesterday evening - having a different primary email address on github makes things difficult.)

cnuke committed:
run: adjust run script, test mkfs by default

cnuke committed:
cbe: remove unused enums

cnuke committed:
lib/cbe: flush

cnuke committed:
cbe: stash flush WIP

cnuke committed:
cbe: stash flush wip 2

cnuke committed:
cbe: remove interal Block_data from WB module

As it is only be used for crypto now anyway (which also is on its way
out...).

cnuke committed:
cbe: make hashing data in WB const

cnuke committed:
cbe: initial reserve handling WIP

cnuke committed:
cbe: reverse handling ff.

cnuke committed:
cbe: intial snapshot handling -- broken

cnuke committed:
cbe: assert objects have the proper size

cnuke committed:
cbe_block: remove snapshot handling DBG messages

cnuke committed:
cbe_block: ignore snapshots w/o height

cnuke committed:
cbe: properly save snapshot

cnuke committed:
cbe: make Primitive printable

cnuke committed:
cbe: add some dbg cxx_ methods to cache module

cnuke committed:
cbe: add dbg macros to I/O module

cnuke committed:
cbe: add ddbg macros to sync SB module

cnuke committed:
cbe: use cache cxx wrapper in free tree

cnuke committed:
cbe: use cache cxx wrapper in VBD

cnuke committed:
cbe: improve dbg messages in WB module

cnuke committed:
cbe: initial SB securing handling

cnuke committed:
cbe: fix missing cache module progress

cnuke committed:
cbe: dbg main loop

and while there remove old code

cnuke committed:
cbe: drop dead code

cnuke committed:
cbe: make Request be printable

cnuke committed:
cbe: replace Genode::error in modules

cnuke committed:
cbe: remove superfluous checks in _data() helper

cnuke committed:
cbe: use cache to determine is syncing is necessary

cnuke committed:
cbe: add progress logging macro

cnuke committed:
cbe: decouple securing SB from snapshot syncing

cnuke committed:
cbe: hook-up discarding of snapshots

cnuke committed:
cbe: add timestamp to macros

cnuke committed:
cbe: normalize dbg timestamp and print in decimal

cnuke committed:
cbe: remove unused modules

cnuke committed:
lib/cbe: initial Cache_Flusher module

cnuke committed:
cbe: hook-up Cache_Flusher module

cnuke committed:
cbe: silence FT dbg messages for now

cnuke committed:
lib/cbe: initial Sync_Superblock module

cnuke committed:
cbe: hook-up Sync_superblock module

cnuke committed:
cbe: add missing Sync superblock header

cnuke committed:
lib/cbe: add missing Sync_superblock files

Commit:
cbe: tree helper, translation, and VBD in SPARK

cnuke committed:
cbe: move event-loop code into Cbe::Library

cnuke committed:
cbe: add cxx wrapper to Crypto module

Primarily to ease debugging.

cnuke committed:
cbe: use Crypto cxx wrapper

cnuke committed:
cbe: more dbg message in Vbd

cnuke committed:
cbe: more dbg in Translation module

cnuke committed:
cbe: simplify Write_back module

cnuke committed:
cbe: rename sealing and securing

And while there, move cache flushing into seal generation scope.

cnuke committed:
cbe: adapt I/O module -- WIP

* use external Io_data → need to change Cache module
  (Job_Data ↔ Io_data)

cnuke committed:
cbe: use external I/O data

(preparatory work for moving Block connection out of CBE)

cnuke committed:
Revert "cbe: use external I/O data"

This reverts commit 2e60cda2fb9d8e511beacfe77a9902ff8842a302

cnuke committed:
cbe: add snapshot slot method to SB

cnuke committed:
cbe: use snapshot slot interface

* start documenting the code

cnuke committed:
cbe: documentation and cleaning up ff

cnuke committed:
cbe: documentation and cleaning up fff

cnuke committed:
cbe: documentation and cleaning up ffff

cnuke committed:
cbe: documentation and cleaning up fffff

cnuke committed:
cbe: document why there is the dummy file

cnuke committed:
cbe: remove old sync super-block module

cnuke committed:
cbe: documentation and cleaning up ffffff

cnuke committed:
cbe: documentation and cleaning up fffffff

cnuke committed:
cbe: documenting free-tree -- WIP

cnuke committed:
cbe: remove use of pointer from FT

cnuke committed:
cbe: fix reading secure interval

cnuke committed:
cbe: document and clean up FT module -- WIP

cnuke committed:
cbe: remove storing pointer in I/O module

cnuke committed:
cbe: move backend I/O out of CBE library

cnuke committed:
cbe: progress in VBD only when submitting cache

cnuke committed:
cbe: remove unused stuff from local util header

cnuke committed:
lib/cbe_cxx: refactor CBE C++ loop into library

cnuke committed:
cbe: use cbe_cxx library

cnuke committed:
lib/cbe_cxx: document FT module

cnuke committed:
lib/cbe_cxx: document Write_back module

cnuke committed:
lib/cbe_cxx: finished documenting I/O module

Commit:
cbe: free tree in SPARK (compiled, not tested)

ssumpf committed:
vfs_cbe: sekeleton

 * dir-file system and block-file system

ssumpf committed:
.gitignore: tmp files (editor, patch)

ssumpf committed:
cbe_vfs: Vfs_handle skeleton

Commit:
cbe: rebase onto cnuke/cbe_wip

Commit:
cbe: spark free_tree (leaf_usable)

Commit:
cbe: free tree (execute)

Commit:
cbe: free tree (peekgenerated primitive)

Commit:
cbe: free tree (Drop_Generated_Primitive)

Commit:
cbe: free tree (Mark_Generated_Primitive_Complete)

nfeske committed:
CBE.Block_IO module

Commit:
cbe: free tree (cxx binding 1)

Commit:
cbe: free tree (cxx binding 2)

nfeske committed:
CBE.Write_Back module

Commit:
cbe: do not use timer in library

Commit:
cbe: move timer impl from library.cc to main.cc

Commit:
cbe: free tree (cxx binding 3)

Commit:
cbe: library (discard_snapshot, compiles)

nfeske committed:
server/cbe: deterministic I/O

This patch moves the notification of the I/O back end to the outside of
the _handle_requests loop, fostering the deterministic behavior of the
CBE across multiple runs.

Commit:
cbe: spark primitive (FIXME in Equal)

Commit:
cbe: spark free tree (remaining subprograms)

Commit:
cbe: spark library (Object_Type, compiles)

nfeske committed:
fixup "cbe: free tree"

Commit:
cbe: add Initialized_Object to all modules

Commit:
cbe: spark library (Initialize_Object, compiles)

Commit:
cbe: spark library (Initialize_Object, timeout requests, compiles)

ssumpf committed:
cbe: add 'valid' to Super_block_index

ssumpf committed:
cbe_vfs: read/write implemenation

ssumpf committed:
cbe_vfs: XML configuration support

Commit:
cbe: c++ library (execute_progress, documentation, succeeds)

Commit:
cbe: spark library (Max_VBA)

Commit:
cbe: c++ library (give_read_data w/o return value)

nfeske committed:
Incorporate Tag in comparison in Primitive.Equal

Commit:
cbe: c++ library (frontend/backend_req_prim)

Commit:
cbe: spark library (Front_End/Back_End_Req_Prim)

nfeske committed:
Simplify Request.Success_Type

This change avoids convoluted conditions like

  Request."=" (Primitive.Success (Prim), Request.False)

nfeske committed:
Use constants for array bounds

nfeske committed:
.gitignore backup files

nfeske committed:
Use Write_Back.New_PBAs_Type in Free_Tree

This commit merges the two types Free_Tree.WB_Data_New_PBAs_Type and
Write_Back.New_PBAs_Type.

nfeske committed:
Add Request.Equal

This function is needed by Library.Give_Write_Data.

nfeske committed:
Library.Give_Write_Data (WiP)

nfeske committed:
Virtual_Block_Device::Trans_Can_Get_Type_1_Info_Spark

Wrapper for Translation::Trans_Can_Get_Type_1_Info_Spark, needed
by Library.Give_Write_Data.

nfeske committed:
fixup Library.Give_Write_Data (done)

nfeske committed:
Library.Need_Data

nfeske committed:
Library.Take_Read_Data

nfeske committed:
Library.Request_Acceptable, Submit_Request ...

nfeske committed:
Library.Ack_Read_Data

Commit:
cbe: spark library (Execute 29%, compiles)

ssumpf committed:
cbe_vfs: run_genode_until timeout

ssumpf committed:
cbe_vfs: debugging

nfeske committed:
Library.Take_Write_Data, Ack_Write_Data, Have_Data

nfeske committed:
Library.Give_Data_Index

nfeske committed:
Library.Give_Read_Data

ssumpf committed:
cbe_vfs: basic functionality (with DEBUGGING)

nfeske committed:
disable strict warnings for vfs_cbe

nfeske committed:
Functional decryption/encryption

The Crypto.Copy_Completed_Data implementation did not implement the flow
of data for plain and cipher text completely. Hence, the actual
encryption and decryption of the CBE was not functional until now.

This patch splits the Crypto.Copy_Completed_Data into two procedures
Copy_Decrypted_Data (for read operations) and Copy_Encrypted_Data (for
write operations) to clear up the flow and buffering of information.

Thanks a lot to @ssumpf for his great support!

Commit:
cbe: spark primitive copy_valid_object_change_tag

Commit:
cbe: spark library (Execute 100%)

Commit:
cbe: library (same interface in c++ and spark)

Commit:
cbe: cbe.cxx.cxx_library (without export clauses)

Commit:
cbe: cbe.cxx.cxx_library (compiles, test fails)

Commit:
cbe: remove no longer used C++-SPARK-transition files

nfeske committed:
server/cbe: increase stack size

nfeske committed:
server/cbe: pass super blocks as const &

nfeske committed:
Library.Execute: Drop prim from cache after access

nfeske committed:
cbe: library translation fixes

I was a bit overzealous with my ambition to reduce code duplications.

nfeske committed:
fixup "cbe: spark library (Execute 100%)"

Apply fix of "Functional decryption/encryption" to the SPARK translation.

nfeske committed:
cbe: add Print_String debug utility

nfeske committed:
cbe: Reflection of internal CBE state

This patch equips the cbe with basic support for obtaining its internal
state in textual form, which allows the tracing of state changes.

nfeske committed:
cbe: fix handling of snapshot indices

nfeske committed:
Streamline C++ library interface

- Remove notion of "Public_Library"
- Move Ada procedure bindings to private section, simplifying
  the use of the public interface.
- Remove stale cbe_cxx library description file

nfeske committed:
cbe_vfs: adaptation to SPARK version of CBE

Commit:
cbe: move timout handling completely to C++

Commit:
cbe: remove deprecated Request.Tag_Type

Commit:
cbe: CBE.Request.Block_Number_Type to CBE

Commit:
cbe: (CXX_)Request.Operation_Type to CBE(.CXX)

Commit:
cbe: split up crypto.submit_primitive

Commit:
cbe_vfs: adapt to recent cbe state

Commit:
cbe: use default stack size (reverts "server/cbe: increase stack size")

Commit:
cbe: do not depend on SHA256_4K in cbe.ads

Commit:
cbe.ads: remove unnecessary size decls and corresponding CXX_* types

Commit:
cbe: streamline cbe-cxx interface

Commit:
cbe: move CXX glue code to dedicated lib

Commit:
server/cbe: remove unnecessary Ada stuff

cnuke committed:
cbe: adapt library interface

cnuke committed:
cbe: temporarily limit request pool to 1 pending request

Otherwise later incoming requests be execute before already stored
once (due to the way the array is managed).

cnuke committed:
cbe: adapt VFS plugin to interface changes

cnuke committed:
cbe: split FT Execute procedure

First step for refactoring the FT.

cnuke committed:
cbe: initial external Crypto interface

cnuke committed:
external_crypto: initial implementation in SPARK

Commit:
cbe_block: write initial state to linux file

Commit:
cbe: restore dummy.ads and call to adainit()

This reverts commit 214ad7a67b2538b165cce779eca48cbc0c4e47da.

cnuke committed:
cbe: re-enable rm -rf libraries in run-script

cnuke committed:
external_crypto: add minimal test

Commit:
sha256_4k: style fixes, license headers

Commit:
cbe: style fixes

cnuke committed:
cbe: use External.Crypto interface

cnuke committed:
external_crypto: add C++ wrapper

cnuke committed:
cbe: hook-up External.Crypto interface in server

Commit:
cbe: provide some aliases for uint types

Commit:
cbe: get rid of unions

Commit:
cbe: name superblock consistently

Commit:
cbe: adapt coding style to External.Crypto

Commit:
server/aes_cbc_block: remove

Commit:
lib/cbe: move crypto buffers to library

Commit:
cbe: move crypto buffers to server main

cnuke committed:
cbe: advance query branch when found free blocks

This fixes a problem (or rather a sympthom) in the FT module where a
query branch without free PBAs is used to populate the WB list.

cnuke committed:
cbe: fix next superblock selection

Apparently, 'mod' comes for '+' (I have not checked the Ada
documentation yet).

Commit:
lib/cbe: replace Crypto buffer copying by indexing

cnuke committed:
cbe: warp snapshot slot properly

Instead of increasing 'Cur_Snap' we have to set it to the calculated
slot (was already wrong in the C++ code).

cnuke committed:
cbe: store current query VBA out-side

Resetting 'Curr_Query_Prim' by assigning 'Invalid_Primitive' also sets
the Block_Number back to 0. The invalid object is then used to
calculate the next VBA, which in this case will always be 64. This
will lead to an endless loop when the FT has to query more than 2
branches to find free blocks.

cnuke committed:
cbe: use compare test in run script

Commit:
cbe: revert initialization of crypto buffers

cnuke committed:
cbe: move Block::Request conversion into server

It is only needed in this case.

Commit:
cbe: obtain_client_data_2

Commit:
cbe: obtain_client_data without crypto buffer

Commit:
cbe: same crypto buffer index range as in C++

Commit:
cbe: replace Crypto Data buffer with index

cnuke committed:
cbe/vfs: force gnatbind

cnuke committed:
server/cbe: move loop counter into handler

cnuke committed:
cbe/vfs: adapt to library interface

* limit to 1 pending request (matches current CBE library setting)
  to prevent reading wrong data after a write

* check request properties (offset, count) and reject invalid ones

cnuke committed:
cbe_block: ignore missing state Report session

cnuke committed:
vfs/cbe: initial directory structure -- WIP

Commit:
server/cbe: get rid of crypto data member

And account for multiple pending crypto request by changing the
while loop.

Commit:
cbe: use index for identification WRT Crypto

Commit:
lib/cbe: static value Block_Size

Commit:
lib/cbe*: remove outcommented code

Commit:
cbe: Key_ID_Type/Key_ID_Storage_Type w/o mod

Commit:
cbe: get rid of invalid snapshot ID

Commit:
lib/cbe: get rid of null statements

Commit:
lib/cbe: get rid of Snapshot_ID_Invalid_Slot

Commit:
lib/cbe: clean up snapshot ID and index

Commit:
lib/cbe: static value Superblock_Nr_Of_Snapshots

Commit:
lib/cbe: get rid of Address_Type and "*_SPARK"

Commit:
cbe: clean up C++ Superblock_index

Commit:
Set up TravisCI

Commit:
Add memlimit

Commit:
cbe: move backend IO buffer out of CBE library

Commit:
lib/cbe: submit IO primitive w/o IO buffer

Commit:
cbe: use data indices for backened IO

Commit:
vfs/cbe: adapt to new CBE crypto API

cnuke committed:
lib/cbe: add snapshot interface

Commit:
cbe: streamline IO backend interface

Commit:
lib/cbe: no Mark_Generated_Primitive_Complete_2

cnuke committed:
lib/cbe: add active snapshots interface

cnuke committed:
server/cbe_block: do not flag initial snapshot

Commit:
lib/cbe: change functions with 'in out' parameters

In particular:

* Cache.Data_Index
* Library.Discard_Snapshot
* Library.Supply_Client_Data
* Write_Back.Peek_Generated_Cahce_Update_PBA
* Crypto.Peek_Generated_Primitive

Commit:
.ci/cbe.gpr: show full paths on gnatprove

Commit:
cbe: proving in mode "check_all" succeeds

When replacing in .ci/prove.mk the gnatprove switch '--checks-as-errors' with
'--mode=check_all', the proving (make -f .ci/prove.mk prove_spark_ci)
succeeds.

Commit:
lib/cbe: make debugging code optional

cnuke committed:
lib/cbe: use current query VBA to check for EOT

Using 'Curr_Query_Prim := Primitive.Invalid_Object' will also reset the
Block_Number that is needed in the check (stems from C++ code where
invalidating was done via setting the Operation to INVALID but keeping
reset of the object intact).

Commit:
cbe: get rid of Obtain_Client_Data_2

This was a dirty way of triggering the submit of an block-IO primitive for
each completed VBD primitive. This is now done internally during the
'CBE.Library.Execute' call.

cnuke committed:
server/cbe_block: increase report buffer size

cnuke committed:
server/cbe_block: show invalid block access

cnuke committed:
lib/cbe: add Active function to Cache_Flusher

For now it is used to check if the flusher is currently doing
something.

cnuke committed:
lib/cbe: make sure to flush cache before securing

cnuke committed:
fixup 'cbe: get rid of invalid snapshot ID'

Set new snapshot to be valid.

cnuke committed:
run/cbe: disarm securing -- XXX

cnuke committed:
lib/cbe: add To_String (Hash)

cnuke committed:
lib/cbe: add Next_Snap_Slot function

The whole approach to selecting the next snapshot slot probably has to
be changed WRT to verification. In there is no snapshot that can be
safely discarded user interventation is needed and therefor it is
possible for 'Cur_Snap = Next_Snap' to occur.

cnuke committed:
lib/cbe: unconditionally updaste snapshot PBA

cnuke committed:
DEBUG -- XX

cnuke committed:
server/cbe_block: mark execute progression

cnuke committed:
lib/cbe: secure superblock on snapshot creation

cnuke committed:
server/cbe: use snapshot interface -- XXX

cnuke committed:
lib/cbe: add snapshot interface

cnuke committed:
fixup 'lib/cbe: add snapshot interface'

cnuke committed:
fixup 'lib/cbe: add snapshot interface' (2)

cnuke committed:
fixup 'vfs/cbe: initial directory structure -- WIP'

cnuke committed:
fixup 'vfs/cbe: initial directory structure -- WIP' (list snap dirs)

cnuke committed:
cbe: export snapshots -- WIP

cnuke committed:
cbe/lib: set flags on snapshot

cnuke committed:
lib/cbe: write-back debug -- XXX

cnuke committed:
fixup 'lib/cbe: add snapshot interface' (3)

- stall Create_Snapshot so that we do not end up creating it in
between write-back operation
